x/shield/keeper/proposal.go

```diff
@@ -1,19 +1,616 @@
 package keeper
 
 import (
+	"fmt"
+	"sort"
 	"time"
 
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	vestexported "github.com/cosmos/cosmos-sdk/x/auth/vesting/exported"
+	"github.com/cosmos/cosmos-sdk/x/staking"
+
+	"github.com/certikfoundation/shentu/x/shield/types"
 )
 
-func (k Keeper) ClaimLock(ctx sdk.Context, proposalID, poolID uint64, purchaser sdk.AccAddress, purchaseID uint64, loss sdk.Coins, lockPeriod time.Duration) error {
+// SecureCollaterals is called after a claim is submitted to secure
+// the given amount of collaterals for the duration and adjust shield
+// module states accordingly.
+func (k Keeper) SecureCollaterals(ctx sdk.Context, poolID uint64, purchaser sdk.AccAddress, purchaseID uint64, loss sdk.Coins, duration time.Duration) error {
+	lossAmt := loss.AmountOf(k.sk.BondDenom(ctx))
+
+	// Verify shield.
+	pool, found := k.GetPool(ctx, poolID)
+	if !found {
+		return types.ErrNoPoolFound
+	}
+	if lossAmt.GT(pool.Shield) {
+		return types.ErrNotEnoughShield
+	}
+
+	// Verify collateral availability.
+	totalCollateral := k.GetTotalCollateral(ctx)
+	totalClaimed := k.GetTotalClaimed(ctx)
+	totalClaimed = totalClaimed.Add(lossAmt)
+	if totalClaimed.GT(totalCollateral) {
+		panic("total claimed surpassed total collateral")
+	}
+
+	// Secure the updated loss ratio from each provider to cover total claimed.
+	providers := k.GetAllProviders(ctx)
+	lossRatio := totalClaimed.ToDec().Quo(totalCollateral.ToDec())
+	for i := range providers {
+		secureAmt := providers[i].Collateral.ToDec().Mul(lossRatio).TruncateInt()
+		if secureAmt.GT(totalClaimed) {
```

Comment:
Move this if to after next if. Otherwise the logic flow seems unnatural.

Reply:
Actually I found these two conditions are not good enough. Will fix it.

Reply:
@dshe-certik Updated. Could you take a look at the updated logic?

Reply:
Looks good.